tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/JSDesugaring.scala

```diff
@@ -2204,21 +2212,37 @@ private[emitter] class JSDesugaring(internalOptions: InternalOptions) {
       implicit outputMode: OutputMode, pos: Position): js.Tree =
     envField("c", className)
 
-  private[emitter] def genRawJSClassConstructor(className: String)(
+  private[emitter] def genRawJSClassConstructor(className: String,
+      enclosingClassName: Option[String])(
       implicit globalKnowledge: GlobalKnowledge, outputMode: OutputMode,
       pos: Position): js.Tree = {
 
-    genRawJSClassConstructor(className,
+    genRawJSClassConstructor(className, enclosingClassName,
         globalKnowledge.getJSNativeLoadSpec(className))
   }
 
   private[emitter] def genRawJSClassConstructor(className: String,
-      spec: Option[JSNativeLoadSpec])(
+      enclosingClassName: Option[String], spec: Option[JSNativeLoadSpec])(
       implicit outputMode: OutputMode, pos: Position): js.Tree = {
     spec match {
       case None =>
-        // this is a Scala.js-defined JS class
-        encodeClassVar(className)
+        /* This is a Scala.js-defined JS class.
+         * Unless it is the enclosing class, it should be referred to using its
+         * class value accessor. Within one class, references to itself
+         * directly refer to the class var, which locally shadows the accessor.
+         */
+        if (enclosingClassName.exists(_ == className)) {
+          encodeClassVar(className)
+        } else {
+          /* In ECMAScript51Global mode, the accessor in under `ScalaJS.a`
+           * rather than `ScalaJS.c` to avoid clashes.
+           */
+          val accessor = outputMode match {
+            case OutputMode.ECMAScript51Global => envField("a", className)
+            case _                             => encodeClassVar(className)
+          }
```

Comment:
Doing so actually significantly improves the situation, as `enclosingClassName` is not necessary anymore, so I can revert all the changes I made about `enclosingClassName`.